File: LeetCode-2023/1814.统计一个数组中好对子的数目.cpp
/*
 * @lc app=leetcode.cn id=1814 lang=cpp
 *
 * [1814] 统计一个数组中好对子的数目
 */

// @lc code=start
class Solution {
public:
    int revFunc(int num) {
        int rev = 0;
        while(num) {
            rev = rev * 10 + num % 10;
            num = num / 10;
        }
        return rev;
    }
    int countNicePairs(vector<int>& nums) {
        int mod = 1e9 + 7;
        long result = 0;
        int n = nums.size();
        unordered_map<int, long> mp;
        for(int i = 0; i < n; i++) {
            mp[nums[i] - revFunc(nums[i])]++;
        }
        for(auto it : mp) {
            result += (it.second * (it.second - 1) / 2) % mod;
        }
        return (int)result % mod;
    }``
};
// @lc code=end

File: LeetCode-2023/1814.统计一个数组中好对子的数目.java
/*
 * @lc app=leetcode.cn id=1814 lang=java
 *
 * [1814] 统计一个数组中好对子的数目
 */

// @lc code=start
class Solution {
    public int revFunc(int num) {
        int rev = 0;
        while(num != 0) {
            rev = rev * 10 + num % 10;
            num /= 10;
        }
        return rev;
    }
    public int countNicePairs(int[] nums) {
        int result = 0;
        int n = nums.length;
        int mod = 1000000007;
        Map<Integer, Integer> mp = new HashMap<>();
        for(int i = 0; i < n; i++) {
            mp.put(nums[i] - revFunc(nums[i]), mp.getOrDefault(nums[i] - revFunc(nums[i]), 0) + 1);
        }
        for(Map.Entry<Integer, Integer> entry : mp.entrySet()) {
            int val = entry.getValue();
            result += ((long)val * (val - 1) / 2) % mod;    // avoid overflow
        }
        return result % mod;
    }
}
// @lc code=end

